server-common/src/main/java/org/apache/kafka/server/mutable/BoundedList.java

```diff
@@ -40,18 +40,17 @@ public static <E> BoundedList<E> newArrayBacked(int maxLength) {
     }
 
     public static <E> BoundedList<E> newArrayBacked(int maxLength, int initialCapacity) {
+        if (initialCapacity <= 0) {
+            throw new IllegalArgumentException("Invalid non-positive initialCapacity of " + initialCapacity);
+        }
         return new BoundedList<>(maxLength, new ArrayList<>(initialCapacity));
     }
 
-    public BoundedList(int maxLength, List<E> underlying) {
+    private BoundedList(int maxLength, List<E> underlying) {
         if (maxLength <= 0) {
             throw new IllegalArgumentException("Invalid non-positive maxLength of " + maxLength);
         }
         this.maxLength = maxLength;
-        if (underlying.size() > maxLength) {
```

Comment:
this check is still useful to me since all helpers are based on this constructor, and so it is a good safety catch.